File: lamblin-integration-test/lamblin-integration-test-java/src/test/java/com/lamblin/it/client/DeleteControllerClient.java
/*
 * Lamblin
 * Copyright 2018 Borislav Shekerov
 * Licensed under Apache 2.0: https://github.com/BorislavShekerov/lamblin/blob/master/LICENSE
 */

package com.lamblin.it.client;

import com.lamblin.it.model.ResponseEntity;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;
import retrofit2.http.DELETE;
import retrofit2.http.Path;
import retrofit2.http.Query;

import static com.lamblin.it.client.ClientUtils.createObjectMapper;
import static com.lamblin.it.client.ClientUtils.executeRequest;
import static com.lamblin.it.model.EndpointsKt.*;
import static com.lamblin.it.model.TestUtilsKt.*;

public final class DeleteControllerClient {

    public static final DeleteControllerClient INSTANCE = new DeleteControllerClient();
    private final DeleteControllerApi client;

    private DeleteControllerClient() {
        Retrofit retrofit = new Retrofit.Builder()
                .addConverterFactory(JacksonConverterFactory.create(createObjectMapper()))
                .baseUrl(getServerBaseUrl())
                .build();

        this.client = retrofit.create(DeleteControllerApi.class);
    }

    public Response<ResponseEntity> callSimpleDeleteNoParamsEndpoint() {
        return executeRequest(client::callSimpleDeleteNoParamsEndpoint);
    }

    public Response<ResponseEntity> callSingleQueryParamEndpoint(String queryParam) {
        return executeRequest(() -> client.callSingleQueryParamEndpoint(queryParam));
    }

    public Response<ResponseEntity> callMultiQueryParamEndpoint(String queryParam1, String queryParam2) {
        return executeRequest(() -> client.callQueryParamEndpoint(queryParam1, queryParam2));
    }

    public Response<ResponseEntity> callSinglePathParamEndpoint(String pathParam) {
        return executeRequest(() -> client.callSinglePathParamEndpoint(pathParam));
    }

    public Response<ResponseEntity> callMultiPathParamEndpoint(String pathParam1, String pathParam2) {
        return executeRequest(() -> client.callMultiPathParamEndpoint(pathParam1, pathParam2));
    }

    public Response<ResponseEntity> callMultiPathParamWithQueryParamEndpoint(
            String queryParam,
            String pathParam1,
            String pathParam2) {

        return executeRequest(() -> client.callMultiPathParamEndpoint(pathParam1, pathParam2, queryParam));
    }

    private interface DeleteControllerApi {

        @DELETE(SIMPLE_DELETE_ENDPOINT)
        Call<ResponseEntity> callSimpleDeleteNoParamsEndpoint();

        @DELETE(QUERY_PARAM_DELETE_ENDPOINT)
        Call<ResponseEntity> callSingleQueryParamEndpoint(@Query(QUERY_PARAM_1) String queryParam);

        @DELETE(QUERY_PARAM_DELETE_ENDPOINT)
        Call<ResponseEntity> callQueryParamEndpoint(@Query(QUERY_PARAM_1) String queryParam1,
                                                    @Query(QUERY_PARAM_2) String queryParam2);

        @DELETE(SINGLE_PATH_PARAM_DELETE_ENDPOINT)
        Call<ResponseEntity> callSinglePathParamEndpoint(@Path(PATH_PARAM_1) String pathParam1);

        @DELETE(MULTIPLE_PATH_PARAM_DELETE_ENDPOINT)
        Call<ResponseEntity> callMultiPathParamEndpoint(@Path(PATH_PARAM_1) String pathParam1,
                                                        @Path(PATH_PARAM_2) String pathParam2);

        @DELETE(MULTIPLE_PATH_PARAM_DELETE_ENDPOINT)
        Call<ResponseEntity> callMultiPathParamEndpoint(@Path(PATH_PARAM_1) String pathParam1,
                                                        @Path(PATH_PARAM_2) String pathParam2,
                                                        @Query(QUERY_PARAM_1) String queryParam1);
    }
}

File: lamblin-integration-test/lamblin-integration-test-kotlin/src/test/kotlin/com/lamblin/it/client/DeleteControllerClient.kt
/*
 * Lamblin
 * Copyright 2018 Borislav Shekerov
 * Licensed under Apache 2.0: https://github.com/BorislavShekerov/lamblin/blob/master/LICENSE
 */

package com.lamblin.it.client

import com.lamblin.it.model.MULTIPLE_PATH_PARAM_DELETE_ENDPOINT
import com.lamblin.it.model.PATH_PARAM_1
import com.lamblin.it.model.PATH_PARAM_2
import com.lamblin.it.model.QUERY_PARAM_1
import com.lamblin.it.model.QUERY_PARAM_2
import com.lamblin.it.model.QUERY_PARAM_DELETE_ENDPOINT
import com.lamblin.it.model.ResponseEntity
import com.lamblin.it.model.SIMPLE_DELETE_ENDPOINT
import com.lamblin.it.model.SINGLE_PATH_PARAM_DELETE_ENDPOINT
import com.lamblin.it.model.getServerBaseUrl
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.jackson.JacksonConverterFactory
import retrofit2.http.DELETE
import retrofit2.http.Path
import retrofit2.http.Query
import java.util.concurrent.TimeUnit

object DeleteControllerClient {

    private val client: DeleteControllerApi

    init {
        val retrofit = Retrofit.Builder()
            .addConverterFactory(JacksonConverterFactory.create(createObjectMapper()))
            .client(
                OkHttpClient.Builder()
                    .readTimeout(1, TimeUnit.HOURS)
                    .build())
            .baseUrl(getServerBaseUrl())
            .build()
        client = retrofit.create(DeleteControllerApi::class.java)
    }

    fun callSimpleDeleteNoParamsEndpoint() = client.callSimpleDeleteNoParamsEndpoint().execute()

    fun callSingleQueryParamEndpoint(queryParam: String) = client.callSingleQueryParamEndpoint(queryParam).execute()

    fun callMultiQueryParamEndpoint(queryParam1: String, queryParam2: String) = client.callQueryParamEndpoint(
        queryParam1, queryParam2
    ).execute()

    fun callSinglePathParamEndpoint(pathParam: String) = client.callSinglePathParamEndpoint(pathParam).execute()

    fun callMultiPathParamEndpoint(pathParam1: String, pathParam2: String) = client.callMultiPathParamEndpoint(
        pathParam1, pathParam2
    ).execute()

    fun callMultiPathParamWithQueryParamEndpoint(
        queryParam: String,
        pathParam1: String,
        pathParam2: String
    ) = client.callMultiPathParamEndpoint(pathParam1, pathParam2, queryParam).execute()

    internal interface DeleteControllerApi {

        @DELETE(SIMPLE_DELETE_ENDPOINT)
        fun callSimpleDeleteNoParamsEndpoint(): Call<ResponseEntity>

        @DELETE(QUERY_PARAM_DELETE_ENDPOINT)
        fun callSingleQueryParamEndpoint(@Query(QUERY_PARAM_1) queryParam: String): Call<ResponseEntity>

        @DELETE(QUERY_PARAM_DELETE_ENDPOINT)
        fun callQueryParamEndpoint(
            @Query(QUERY_PARAM_1) queryParam1: String,
            @Query(QUERY_PARAM_2) queryParam2: String
        ): Call<ResponseEntity>

        @DELETE(SINGLE_PATH_PARAM_DELETE_ENDPOINT)
        fun callSinglePathParamEndpoint(@Path(PATH_PARAM_1) pathParam1: String): Call<ResponseEntity>

        @DELETE(MULTIPLE_PATH_PARAM_DELETE_ENDPOINT)
        fun callMultiPathParamEndpoint(
            @Path(PATH_PARAM_1) pathParam1: String,
            @Path(PATH_PARAM_2) pathParam2: String
        ): Call<ResponseEntity>

        @DELETE(MULTIPLE_PATH_PARAM_DELETE_ENDPOINT)
        fun callMultiPathParamEndpoint(
            @Path(PATH_PARAM_1) pathParam1: String,
            @Path(PATH_PARAM_2) pathParam2: String,
            @Query(QUERY_PARAM_1) queryParam1: String
        ): Call<ResponseEntity>
    }
}